agent/consul/state/kvs_test.go

```diff
@@ -1022,6 +1024,163 @@ func TestStateStore_KVSDeleteTree(t *testing.T) {
 	}
 }
 
+func TestStateStore_Watches_PrefixDelete(t *testing.T) {
+	s := testStateStore(t)
+
+	// Create some KVS entries
+	testSetKey(t, s, 1, "foo", "foo")
+	testSetKey(t, s, 2, "foo/bar", "bar")
+	testSetKey(t, s, 3, "foo/bar/zip", "zip")
+	testSetKey(t, s, 4, "foo/bar/zip/zorp", "zorp")
+	testSetKey(t, s, 5, "foo/bar/baz", "baz")
+
+	// Delete a key and make sure the index comes from the tombstone.
+	ws := memdb.NewWatchSet()
+	idx, _, err := s.KVSList(ws, "foo/bar/baz")
+	if err != nil {
+		t.Fatalf("unexpected err: %s", err)
+	}
+	if err := s.KVSDeleteTree(6, "foo/bar"); err != nil {
+		t.Fatalf("unexpected err: %s", err)
+	}
+	if !watchFired(ws) {
+		t.Fatalf("expected watch to fire but it did not")
+	}
+	ws = memdb.NewWatchSet()
+	idx, _, err = s.KVSList(ws, "foo/bar/baz")
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+	if idx != 6 {
+		t.Fatalf("bad index: %d, expected %d", idx, 6)
+	}
+
+	// Set a different key to bump the index. This shouldn't fire the
+	// watch since there's a different prefix.
+	testSetKey(t, s, 7, "some/other/key", "")
+	if watchFired(ws) {
+		t.Fatalf("bad")
+	}
+
+	// Make sure we get the right index from the tombstone for the prefix
+	idx, _, err = s.KVSList(nil, "foo/bar")
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	if idx != 6 {
+		t.Fatalf("bad index: %d, expected %v", idx, 7)
+	}
+
+	// Now ask for the index for a node within the prefix that was deleted
+	// We expect to get the max index in the tree because the tombstone contains the parent foo/bar
+	idx, _, err = s.KVSList(nil, "foo/bar/baz")
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	if idx != 7 {
+		t.Fatalf("bad index: %d, expected %v", idx, 7)
+	}
+
+	// Now reap the tombstones and make sure we get the latest index
+	// since there are no matching keys.
+	if err := s.ReapTombstones(6); err != nil {
+		t.Fatalf("err: %s", err)
+	}
+	idx, _, err = s.KVSList(nil, "foo/bar/baz")
```

Comment:
If I implement the other detailed scenario you mentioned above, I think I want to delete the lower part of this test because it is redundant. I will leave out just the part where we insert another unrelated node.